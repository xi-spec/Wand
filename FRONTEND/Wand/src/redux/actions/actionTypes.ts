export default {
  LOAD_PRODUCT_TYPE: 'LOAD_PRODUCT_TYPE',
  LOAD_ALL_PRODUCTS: 'LOAD_ALL_PRODUCTS',
  LOAD_NEW_PRODUCTS: 'LOAD_NEW_PRODUCTS',
  LOAD_PRODUCTS_LIST: 'LOAD_PRODUCTS_LIST',
  LOAD_PRODUCT_DETAIL: 'LOAD_PRODUCT_DETAIL',
  ADD_PRODUCT_TO_CART: ' ADD_PRODUCT_TO_CART',
  INCREASE_QUANTITY_CART: 'INCREASE_QUANTITY_CART',
  DECREASE_QUANTITY_CART: 'DECREASE_QUANTITY_CART',
  DELETE_PRODUCT_CART: 'DELETE_PRODUCT_CART',
  DELETE_ALL_CART: 'DELETE_ALL_CART',
  UPDATE_STOCK_API: 'UPDATE_STOCK_API',
  SEARCH_PRODUCT: 'SEARCH_PRODUCT',
  ADD_PRODUCT_TO_FAVOURITE: 'ADD_PRODUCT_TO_FAVOURITE',
  REGISTER: 'REGISTER',
  REGISTER_ERROR: 'REGISTER_ERROR',
  LOGIN: 'LOGIN',
  LOGIN_ERROR: 'LOGIN_ERROR',
  CLEAN_ISEXIST_STATUS: 'CLEAN_STATUS',
  CLEAN_LOGIN_ERROR: 'CLEAN_LOGIN_ERROR',
  CLEAN_REGISTED: 'CLEAN_REGISTED',
  REMOVE_FAVOURITE: 'REMOVE_FAVOURITE',
  LOGOUT: 'LOGOUT',
  UPDATE_USER: 'UPDATE_USER',
  PAYMENT_SUCCSSFULY: 'PAYMENT_SUCCESSFULY',
  PDATE_STOCK_API: 'PDATE_STOCK_API'
}
;
