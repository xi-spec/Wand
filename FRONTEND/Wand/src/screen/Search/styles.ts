import { StyleSheet } from 'react-native';
import { FONTS } from '../../constants/theme';

const styles = StyleSheet.create({
  container: {
    height: '100%'
  },
  searchInput: {
    fontFamily: FONTS.primary
  }

});

export default styles;
