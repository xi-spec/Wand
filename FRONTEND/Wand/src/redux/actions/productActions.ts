import actionTypes from './actionTypes';
import axios from 'axios';
import SERVER from '../../constants/server';
import { Dispatch } from 'redux';
import { Product, ShoppingCart } from '../../models/initiaState';

export function loadProductsType () {
  return async (dispatch : Dispatch) => {
    const { data } = await axios.get(SERVER.URL_TYPE);
    dispatch({
      type: actionTypes.LOAD_PRODUCT_TYPE,
      data
    });
  };
}

export function loadAllProducts () {
  return async (dispatch : Dispatch) => {
    const { data } = await axios.get(SERVER.URL_PRODUCT);
    dispatch({
      type: actionTypes.LOAD_ALL_PRODUCTS,
      data
    });
  };
}

export function loadProductsList (typeName:string) {
  return {
    type: actionTypes.LOAD_PRODUCTS_LIST,
    typeName
  };
}

export function loadProductDetail (productId:string) {
  return {
    type: actionTypes.LOAD_PRODUCT_DETAIL,
    productId
  };
}

export function addProductToCart (product:Product) {
  return {
    type: actionTypes.ADD_PRODUCT_TO_CART,
    product
  };
}

export function increaseQuantityCart (productId:string) {
  return {
    type: actionTypes.INCREASE_QUANTITY_CART,
    productId
  };
}

export function decreaseQuantityCart (productId:string) {
  return {
    type: actionTypes.DECREASE_QUANTITY_CART,
    productId
  };
}

export function deleteProductCart (productId:string, quantity:number) {
  return {
    type: actionTypes.DELETE_PRODUCT_CART,
    productId,
    quantity
  };
}

export function searchProducts (name:string) {
  return {
    type: actionTypes.SEARCH_PRODUCT,
    name
  };
}

export function updateStockApi (cart:ShoppingCart) {
  return (dispatch : Dispatch) => {
    cart.cartList.forEach(async (product) => {
      const { data } = await axios.put(`${SERVER.URL_PRODUCT}/${product._id}`, { stock: product.stock });
      dispatch({
        type: actionTypes.UPDATE_STOCK_API,
        data
      });
    });
  };
}
