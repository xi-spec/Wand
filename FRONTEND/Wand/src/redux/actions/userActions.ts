import axios from 'axios';
import actionTypes from './actionTypes';
import SERVER from '../../constants/server';
import { User, Product, ShoppingCart } from '../../models/initiaState';
import { Dispatch } from 'redux';

export function addProductToFavourite (product:Product, user:User, favouriteState:boolean) {
  let newUser;
  favouriteState
    ? newUser = { ...user, favourite: [...user.favourite, product] }
    : newUser = { ...user, favourite: user.favourite.filter(elem => elem._id !== product._id) };

  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, newUser);
    dispatch({
      type: actionTypes.ADD_PRODUCT_TO_FAVOURITE,
      data
    });
  };
}

export function register (email:string, name:string, password:string) {
  const query = { email, name, password };
  let actionObj;
  return async (dispatch:Dispatch) => {
    try {
      const { data, status } = await axios.post(SERVER.URL_REGISTER, query);
      actionObj = {
        type: actionTypes.REGISTER,
        data,
        status
      };
    } catch (error) {
      actionObj = {
        type: actionTypes.REGISTER_ERROR
      };
    }
    dispatch(
      actionObj
    );
  };
}

export function login (email:string, password:string) {
  const query = { email, password };

  return async (dispatch:Dispatch) => {
    let actionObj;
    try {
      const { data } = await axios.post(SERVER.URL_LOGIN, query);
      actionObj = {
        type: actionTypes.LOGIN,
        data

      };
    } catch (error) {
      actionObj = {
        type: actionTypes.LOGIN_ERROR
      };
    }

    dispatch(actionObj);
  };
}

export function logout (email:string, password:string) {
  const query = { email, password };

  return async (dispatch:Dispatch) => {
    const { data } = await axios.post(SERVER.URL_LOGOUT, query);

    dispatch({
      type: actionTypes.LOGOUT,
      data
    });
  };
}

export function cleanIsExistStatus () {
  return {
    type: actionTypes.CLEAN_ISEXIST_STATUS
  };
}

export function cleanLoginError () {
  return {
    type: actionTypes.CLEAN_LOGIN_ERROR
  };
}

export function cleanRegisted () {
  return {
    type: actionTypes.CLEAN_REGISTED
  };
}

export function removeFavourite (product:Product, user:User) {
  const newUser = { ...user, favourite: user.favourite.filter(elem => elem._id !== product._id) };

  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, newUser);
    dispatch({
      type: actionTypes.REMOVE_FAVOURITE,
      data
    });
  };
}

export function updateAddress (street:string,
  city:string, country:string, postcode:string, user:User) {
  const query = { street, city, country, postcode };
  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, { address: query });
    dispatch({
      type: actionTypes.UPDATE_USER,
      data
    });
  };
}

export function updateCard (cardNumber:string, securtyCode:string, cardholder:string, expiryDate:string, user:User) {
  const query = { cardNumber, securtyCode, cardholder, expiryDate };
  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, { card: query });
    dispatch({
      type: actionTypes.UPDATE_USER,
      data
    });
  };
}

export function paymentSuccessfuly (user:User, cart:ShoppingCart) {
  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, { orderHistory: [...user.orderHistory, cart] });
    dispatch({
      type: actionTypes.PAYMENT_SUCCSSFULY,
      data
    });
  };
}
