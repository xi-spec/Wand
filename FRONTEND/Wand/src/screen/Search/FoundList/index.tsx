import React from 'react';
import { Product } from '../../../models/initiaState';
import { Text, TouchableOpacity, FlatList, Image } from 'react-native';
import styles from './styles';
import setCostValue2Decimals from '../../../utils/setCostValue2Decimals';

export default function FoundList ({ navigation, foundList }:{navigation:any, foundList:Product[]}) {
  return (
<FlatList
        bounces={false}
        style={styles.listContainer}
        data={foundList}
        numColumns={2}
        keyExtractor={(item) => item._id}
        renderItem={({ item }) =>
        <TouchableOpacity
        style={styles.imageContainer}
        key={item.name}
        testID='foundListImage'
        onPress={() => navigation.navigate('Detail', { productId: item._id })}>
      <Image source={{ uri: item.images.cover }}
      style={styles.image}>
        </Image>
        <Text style={styles.detail}>{item.name}</Text>
        <Text style={styles.cost}>{`${setCostValue2Decimals(item.cost.value)} ${item.cost.currency}`}</Text>
      </TouchableOpacity>
        }>
      </FlatList>
  );
}
