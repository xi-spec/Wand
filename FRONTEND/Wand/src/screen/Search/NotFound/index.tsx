import React from 'react';
import { View, Text, Image } from 'react-native';
import styles from './styles';
import { IMAGES } from '../../../constants/images';
import { NOT_FOUND } from '../../../constants/screenTexts';

export default function NotFound ({ navigation }:any) {
  return (
        <View style={styles.container}>
          <Image style={styles.image}source={{ uri: IMAGES.notFound }}></Image>
        <Text style={styles.text}>{NOT_FOUND.NOT_FOUND_TITLE}</Text>
        <Text style={styles.text}>{NOT_FOUND.NOT_FOUND_SUBTITLE}</Text>
      </View>

  );
}
