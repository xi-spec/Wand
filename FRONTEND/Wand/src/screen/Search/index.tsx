import React, { useEffect, useState } from 'react';
import NotFound from './NotFound';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { View } from 'react-native';
import { SEARCH_TEXT } from '../../constants/screenTexts';
import styles from './styles';
import { Input, Icon } from 'react-native-elements';
import { searchProducts } from '../../redux/actions/productActions';
import FoundList from './FoundList';
import { FilterProducts } from '../../models/initiaState';

function Search ({ navigation, actions, filterProducts }:{navigation:{navigate:Function}, actions:any, filterProducts:FilterProducts}) {
  const [search, setSearch] = useState('');

  useEffect(() => {
    const serachTiming = setTimeout(() => actions.searchProducts(search), 350);
    return () => clearTimeout(serachTiming);
  }, [search]);

  return (
    <View style={styles.container}>

<View>
      <Input
      clearButtonMode='always'
      testID='searchInput'
      inputStyle={styles.searchInput}
      autoFocus={true}
      placeholder={SEARCH_TEXT.SEARCH_INPUT}
      onChangeText={text => setSearch(text)}
      value={search}
      leftIcon={<Icon name="search" type="evilicon" color='grey'/>}>
        </ Input>
</View>
{filterProducts.notFound ? <NotFound /> : <FoundList navigation={navigation} foundList ={filterProducts.foundList} /> }

</View>

  );
}

function mapStateToProps ({ products: { filterProducts } }) {
  return {
    filterProducts
  };
}

function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators({
      searchProducts
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Search);
