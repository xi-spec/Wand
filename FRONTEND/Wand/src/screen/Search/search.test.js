import { render, fireEvent, waitFor } from '@testing-library/react-native';
import React from 'react';
import ConfigureStore from 'redux-mock-store';
import * as actions from '../../redux/actions/productActions';
import { Provider } from 'react-redux';
import Search from './index';

describe('Given a Search component', () => {
  const mockStore = ConfigureStore();
  const store = mockStore({
    products: {
      filterProducts: {
        foundList: [
        ]
      }
    }
  });

  beforeEach(() => {
    jest.spyOn(actions, 'searchProducts').mockReturnValueOnce({ type: '' });
  });
  describe('When is rendered', () => {
    test('Then is matched to snapshop', () => {
      const tree = render(<Provider store={store}><Search /></Provider>);

      expect(tree).toMatchSnapshot();
    });
  });

  describe('When is rendered and notFound prop is true', () => {
    test('Then <NotFound/> component is  matched to snapshop', () => {
      const store = mockStore({
        products: {
          filterProducts: {
            foundList: [

            ],
            notFound: true
          }
        }
      });
      const tree = render(<Provider store={store}><Search /></Provider>);

      expect(tree).toMatchSnapshot();
    });
  });

  describe('When text of search input is change with "HELLO" ', () => {
    test('Then text of search input  to be "HELLO"', () => {
      const { getByTestId } = render(<Provider store={store}><Search /></Provider>);
      const searchInput = getByTestId('searchInput');
      const text = 'HELLO';

      fireEvent.changeText(searchInput, text);

      expect(searchInput.props.value).toBe(text);
    });
  });

  describe('When useEffect is invoked', () => {
    test('Then actions.searchProducts is invoked', async () => {
      render(<Provider store={store}><Search /></Provider>);

      await waitFor(() => expect(actions.searchProducts).toHaveBeenCalled());
    });
  });
});
