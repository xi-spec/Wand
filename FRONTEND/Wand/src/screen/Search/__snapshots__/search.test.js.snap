// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Search component When is rendered Then is matched to snapshop 1`] = `
<View
  style={
    Object {
      "height": "100%",
    }
  }
>
  <View>
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
        <TextInput
          allowFontScaling={true}
          autoFocus={true}
          clearButtonMode="always"
          editable={true}
          onChangeText={[Function]}
          placeholder="What are you searching for?"
          placeholderTextColor="#86939e"
          rejectResponderTermination={true}
          style={
            Object {
              "alignSelf": "center",
              "color": "#242424",
              "flex": 1,
              "fontFamily": "MontserratRegular",
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="searchInput"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "margin": 5,
          }
        }
      />
    </View>
  </View>
  <RCTScrollView
    bounces={false}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "margin": 10,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;

exports[`Given a Search component When is rendered and notFound prop is true Then <NotFound/> component is  matched to snapshop 1`] = `
<View
  style={
    Object {
      "height": "100%",
    }
  }
>
  <View>
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "grey",
                        "fontSize": 24,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "evilicons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="iconIcon"
                >
                  ï„¹
                </Text>
              </View>
            </View>
          </View>
        </View>
        <TextInput
          allowFontScaling={true}
          autoFocus={true}
          clearButtonMode="always"
          editable={true}
          onChangeText={[Function]}
          placeholder="What are you searching for?"
          placeholderTextColor="#86939e"
          rejectResponderTermination={true}
          style={
            Object {
              "alignSelf": "center",
              "color": "#242424",
              "flex": 1,
              "fontFamily": "MontserratRegular",
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="searchInput"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "margin": 5,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "paddingTop": 266.8,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://trello-attachments.s3.amazonaws.com/603be1b68ce9c654d30ae543/60506d7b3b22568450a0dc85/1907e2c9a4ad2ba8986ad1530a402b84/final2-400x400.jpg",
        }
      }
      style={
        Object {
          "borderRadius": 50,
          "height": 100,
          "width": 100,
        }
      }
    />
    <Text
      style={
        Object {
          "fontFamily": "MontserratRegular",
          "fontSize": 14,
          "marginHorizontal": 20,
          "paddingVertical": 20,
        }
      }
    >
      Sorry,we could not find any products
    </Text>
    <Text
      style={
        Object {
          "fontFamily": "MontserratRegular",
          "fontSize": 14,
          "marginHorizontal": 20,
          "paddingVertical": 20,
        }
      }
    >
      
    </Text>
  </View>
</View>
`;
